src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -0,0 +1,462 @@
+            {
+                _buildManager.OnDesignTimeOutputDeleted(item);
+            }
+        }
+
+        /// <summary>
+        /// InitializeInnerCoreAsync is responsible for setting an initial AppliedValue. This value will be used by any UI thread calls that may happen
+        /// before the first data flow blocks have been processed. If this method doesn't return a value then the system will block until the first blocks
+        /// have been applied but since we are initialized on the UI thread, that blocks for us, so we must return an initial state.
+        /// </summary>
+        protected override Task InitializeInnerCoreAsync(CancellationToken cancellationToken)
+        {
+            AppliedValue = new ProjectVersionedValue<DesignTimeInputsItem>(new DesignTimeInputsItem(), ImmutableDictionary.Create<NamedIdentity, IComparable>());
+
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        /// This method is where we tell data flow which blocks we're interested in receiving updates for
+        /// </summary>
+        protected override IDisposable LinkExternalInput(ITargetBlock<IProjectVersionedValue<InputTuple>> targetBlock)
+        {
+            return ProjectDataSources.SyncLinkTo(
+                _projectSubscriptionService.ProjectSource.SourceBlock.SyncLinkOptions(),
+                _projectSubscriptionService.SourceItemsRuleSource.SourceBlock.SyncLinkOptions(),
+                targetBlock,
+                new DataflowLinkOptions { PropagateCompletion = true },
+                cancellationToken: ProjectAsynchronousTasksService.UnloadCancellationToken);
+        }
+
+        /// <summary>
+        /// Preprocess gets called as each data flow block updates and its job is to take the input from those blocks and do whatever work needed
+        /// so that ApplyAsync has all of the info it needs to do its job.
+        /// </summary>
+        protected override Task<DesignTimeInputsDelta> PreprocessAsync(IProjectVersionedValue<InputTuple> input, DesignTimeInputsDelta previousOutput)
+        {
+            ProjectInstance project = input.Value.Item1.ProjectInstance;
+            IProjectChangeDescription changes = input.Value.Item2.ProjectChanges[Compile.SchemaName];
+
+            ImmutableArray<string>.Builder addedDesignTimeInputs = ImmutableArray.CreateBuilder<string>();
+            ImmutableArray<string>.Builder removedDesignTimeInputs = ImmutableArray.CreateBuilder<string>();
+            ImmutableArray<string>.Builder addedDesignTimeSharedInputs = ImmutableArray.CreateBuilder<string>();
+
+            foreach (string item in changes.Difference.AddedItems)
+            {
+                PreprocessAddItem(item);
+            }
+
+            foreach (string item in changes.Difference.RemovedItems)
+            {
+                // the RemovedItems doesn't have any project info any more so we can't tell if an item was a design time input or not
```

Comment:
It does; IProjectChangeDescription.Before.Items contains metadata.

Reply:
This has been resolved by changes due to the suggestion ProjectRuleSource hint in https://github.com/dotnet/project-system/pull/4349#discussion_r238940967